contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,905 @@
+import {IIntegrationRegistry} from "../interfaces/IIntegrationRegistry.sol";
+import {IWrapV2Adapter} from "../interfaces/IWrapV2Adapter.sol";
+import {IDebtIssuanceModule} from "../interfaces/IDebtIssuanceModule.sol";
+import {ISetToken} from "../interfaces/ISetToken.sol";
+import {IWETH} from "../interfaces/IWETH.sol";
+
+import {Withdrawable} from "external/contracts/aaveV2/utils/Withdrawable.sol";
+import {DEXAdapter} from "./DEXAdapter.sol";
 
 /**
  * @title FlashMintWrapped
  *
- * Flash issues SetTokens whose components are purely made up of wrapped tokens.
- * In particular, for issuance, the contract needs to:
- * 1. For all components in a SetToken, swap input token into matching unwrapped component
- * 2. For each unwrapped component, execute the wrapping function
- * 3. Call on issuanceModule to issue tokens
+ * Flash issues SetTokens whose components contain wrapped tokens.
+ *
+ * Compatible with:
+ * IssuanceModules: DebtIssuanceModule, DebtIssuanceModuleV2
+ * WrapAdapters: IWrapV2Adapter
+ *
+ * Supports flash minting for sets that contain both unwrapped and wrapped components.
+ * Does not support debt positions on Set token.
+ * Wrapping / Unwrapping is skipped for a component if ComponentSwapData[component_index].underlyingERC20 address == set component address
+ * If the set contains both wrapped and unwrapped version of a token (e.g. DAI and cDAI) -> supply two separate component data points
+ * e.g. for issue
+ * Set components at index 0 = DAI; then -> ComponentSwapData[0].underlyingERC20 = DAI; (no wrapping will happen)
+ * Set components at index 1 = cDAI; then -> ComponentSwapData[1].underlyingERC20 = DAI; (wrapping will happen)
  */
-contract FlashMintWrapped is ReentrancyGuard {
-
-    using Address for address payable;
-    using SafeMath for uint256;
-    using PreciseUnitMath for uint256;
-    using SafeERC20 for IERC20;
-    using SafeERC20 for ISetToken;
-
-    struct WrappedCalldata {
-        address target;
-        uint256 quantity;
-        bytes callData;
-    }
+contract FlashMintWrapped is ReentrancyGuard, Withdrawable {
+  using DEXAdapter for DEXAdapter.Addresses;
+  using Address for address payable;
+  using Address for address;
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+  using SafeERC20 for ISetToken;
 
-    /* ============ Enums ============ */
+  /* ============ Structs ============ */
 
-    /* ============ Constants ============= */
+  struct ComponentSwapData {
+    // unwrapped token version, e.g. DAI
+    address underlyingERC20;
+    // swap data for DEX operation: fees, path, etc. see DEXAdapter.SwapData
+    DEXAdapter.SwapData dexData;
+    // ONLY relevant for issue, not used for redeem:
+    // amount that has to be bought of the unwrapped token version (to cover required wrapped component amounts for issuance)
+    // this amount has to be computed beforehand through the exchange rate of wrapped Component <> unwrappedComponent
```

Comment:
If WrapAdapters would include a getter for this exchange rate, we could do this in a more elegant way without requiring the buy amounts of the Underlying (unwrapped) token.... 

Compound, Aave, [...] all should support such functionality. 

We could do this on our own Set fork (i.e. WrapAdapterV3), but that would mean yet another change to original contract code right from the start

Reply:
Yeah i'm thinking with all the logic around exchange rates and the prices on the initial swap to/from WETH, can we push that calculation to the frontend? such that we have the view method take in all the quantities etc and just spit out the relevant calldata that can be passed back into the mint/redeem function?